File: oj/2.cpp
#include <iostream>
#include <string>

class Position {
public:
    int left;
    int right;

    Position(int left, int right) {
        this->left = left;
        this->right = right;
    }
};

Position solve(int *arr, int target, int l, int r) {
    if (r < 0) {
        return Position(-1, -1);
    }

    // conquer
    if (l == r) {
        // leaf node position
        return arr[r] == target ? Position(r, r) : Position(-1, -1);
    }

    // divide
    int m = (l + r) >> 1;
    Position left_result = arr[m] >= target ? solve(arr, target, l, m) : Position(-1, -1);
    Position right_result = arr[m + 1] <= target ? solve(arr, target, m + 1, r) : Position(-1, -1);

    // combine
    if (left_result.left == -1) {
        // in this case, left half does not contain target element.
        return right_result;
    } else {
        // in this case, left part contains at least one target element.
        return right_result.left == -1 ? left_result : Position(left_result.left, right_result.right);
    }
}

int main() {
    std::string line, str;

    // line 1
    int n, m;
    std::cin >> n;
    std::cin >> m;

    // line 2
    int *arr = new int[n];
    for (int i = 0; i < n; ++i) {
        std::cin >> arr[i];
    }

    // targets
    for (int i = 0; i < m; ++i) {
        int target;
        std::cin >> target;
        Position result = solve(arr, target, 0, n - 1);
        std::cout << result.left << " " << result.right << std::endl;
    }
}

File: oj/2.py
# author: ShotgunRO

# ---------- [ subject ] ----------
# [Description]
# Given an array of integers numbers sorted in ascending order, find the starting and ending position of several
# given target values. If a target is not found in the array, return [-1, -1]. You must write an algorithm with
# O(logn) runtime complexity for each given target.
# [Input]
# Line 1: two integers n and m(n >= 1, m <= 10^5) -- the length of array and the number of targets.
# Line 2: the all elements in array and split by spaces(for each element x, -2^31 <= x <= 2^31 - 1)
# Line 3 ~ Line m+2: targets
# [Output]
# For each target, print one line containing the starting and ending position split by spaces.

# ---------- [ code ] ----------
def solve(arr, target, l, r):
    if r < 0:
        return [-1, -1]

    # Conquer
    if l == r:
        # leaf node position
        return [r, r] if arr[r] == target else [-1, -1]

    # Divides
    m = (l + r) >> 1
    left = solve(arr, target, l, m) if arr[m] >= target else [-1, -1]
    right = solve(arr, target, m + 1, r) if arr[m + 1] <= target else [-1, -1]

    # Combine
    if left[0] == -1:
        # in this case, left half does not contain target element.
        return right
    else:
        # in this case, left part contains at least one target element.
        return left if right[0] == -1 else [left[0], right[1]]


n, m = map(int, input().split(' '))
arr = list(map(int, input().split(' ')))
for i in range(m):
    target = int(input())
    result = solve(arr, target, 0, len(arr) - 1)
    print(result[0], result[1])
